File: 0437-path-sum-iii/0437-path-sum-iii.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def preorder(self, node: Optional[TreeNode], curr_sum: int) -> None:
        if not node:
            return
        
        curr_sum += node.val
        
        if curr_sum == self.targetSum:
            self.count += 1
        
        self.count += self.h[curr_sum - self.targetSum]
        self.h[curr_sum] += 1
        
        self.preorder(node.left, curr_sum)
        self.preorder(node.right, curr_sum)
        
        self.h[curr_sum] -= 1
    
    def pathSum(self, root: Optional[TreeNode], targetSum: int) -> int:
        self.count, self.targetSum = 0, targetSum
        self.h = defaultdict(int)
        self.preorder(root, 0)
        return self.count
        

File: 0437-path-sum-iii/0437-path-sum-iii.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
#define ll long long

class Solution {
public:
    int pathSum(TreeNode* root, int targetSum) {
        if (!root) return 0;
        int ans = 0;
        unordered_map<ll, int> h;
        stack<pair<TreeNode*, pair<ll, bool>>> stack;
        stack.push({root, {0, false}});
        while(!stack.empty()){
            auto [node, others] = stack.top();
            auto [curr_sum, visited] = others;
            stack.pop();
            
            if (visited){
                --h[curr_sum];
                continue;
            }
            
            curr_sum += node -> val;
            if (curr_sum == targetSum) ++ans;
            ans += h[curr_sum - targetSum];
            ++h[curr_sum];
            stack.push({node, {curr_sum, true}});
            
            if (node -> right) stack.push({node -> right, {curr_sum, false}});
            if (node -> left) stack.push({node -> left, {curr_sum, false}});
        }
        
        return ans;
    }
};